indico_zodbimport/modules/event_papers.py

```diff
@@ -182,17 +306,80 @@ def _migrate_papers(self):
                 continue
             contribution = contributions[contrib_id]
             self._migrate_paper_roles(contribution)
+            self._migrate_revisions(contribution, rm)
+
+    def _migrate_paper_files(self):
+        n = 0
+        # we look for paper files in increasing order, so that we can detect file reuse
+        q = (LegacyPaperFile.query
+             .filter(Contribution.event_new == self.event)
+             .join(Contribution)
+             .order_by(Contribution.id, LegacyPaperFile.revision_id))
+
+        for contribution_id, revision in itertools.groupby(q, attrgetter('contribution_id')):
+            for revision_id, files in itertools.groupby(revision, attrgetter('revision_id')):
+                ignored_checksums = set()
+                checksum_map = {}
+                if revision_id is not None:
+                    revision = self.legacy_contrib_revision_map[(contribution_id, revision_id)]
+                else:
+                    revision = None
+
+                for lpf in files:
+                    # check whether the same file has been uploaded to a subsequent revision
+                    try:
+                        with lpf.open() as f:
+                            checksum = crc32(f.read())
+                        checksum_map[checksum] = lpf
+                        collision = self.file_checksums.get(checksum)
+                        if collision:
+                            ignored_checksums.add(checksum)
+                            self.importer.print_warning(
+                                cformat('%{yellow!}File {} (rev. {}) already in revision {}').format(
+                                    lpf.filename, revision.id if revision else None, collision.id),
+                                event_id=self.event.id)
+                            continue
+                        else:
+                            self.file_checksums[checksum] = revision
+                    except (RuntimeError, StorageError):
+                        self.importer.print_error(cformat("%{red!}File not accessible; can't CRC it [{}]")
+                                                  .format(lpf.filename), event_id=self.event.id)
+
+                    if revision:
+                        revision.files.append(_paper_file_from_legacy(lpf))
+                    n += 1
+
+                # if a revision has no files (because they've all been ignored), then keep around a copy of each
```

Comment:
To avoid that we have revisions with zero files. Given the bad validation in past versions, it might happen that we have the same file in 3 different revisions (and nothing else). The code CRC-based checking mechanism would only preserve it in the first one.